### DISCUSSION TITLE
Quantos espaços vocês utilizam p/ indentação?

### DISCUSSION THREAD

**Discussion by felipefialho** [frontendbr/forum#1158] [Discussão]
Vamos na polêmica 😛

Quantos espaços vocês utilizam para indentação de código?

[Enquete no Twitter](https://twitter.com/LFeh/status/1018955145630101505)

- 2 Espaços
- 4 Espaços
 
Por que?

**Comment by helderberto** [comment]
Utilizo 2 espaços por gosto particular e o time em que trabalho também, acredito que esses são os principais motivos. 💯

**Comment by t0ledoj** [comment]
Tu quer dizer tab size 2/4 espaços, ou de fato 2/4 espaços? =o

Meu tabsize é 4. Utilizo para manter o padrão do projeto.

**Comment by felipefialho** [comment]
Respondendo eu mesmo:

Utilizo 2 espaços, acho que fica mais fácil de ler o código. Particularmente me incomoda linhas muito espaçadas.

**Comment by eduardomarino** [comment]
Uso 2 espaços, acho que visualmente compõe melhor o código. Outro motivo é que procuro não ultrapassar o limite de 80 caracteres por linha, logo uma indentação maior acabará me fazendo usar mais quebras do que eu gostaria.

**Comment by angelosds** [comment]
Sempre usei 4 espaços, mas ultimamente tenho visto algumas vantagens em usar 2, por isso, comecei a usar em projetos pessoais.

No trampo o padrão é 4 e o pessoal meio que odeia 2 espaços, um dia tento convencê-los. Haha

**Comment by kivervinicius** [comment]
4 espaços, mania de programador Delphi, não consigo visualizar código com 2 espaços para mim fica tudo misturado.

**Comment by renanlara** [comment]
Mas não é justamente para isso que existe o https://editorconfig.org/?

**Comment by kavalcante** [comment]
Eu sigo o Airbnb Styleguide, lá é definido por 2 espaços. E esse lance é mais estético e manter um padrão do que propriamente por funcionamento. Então é isso.

**Comment by diemoritat** [comment]
2 espaços. Primeiro pq eu tento seguir o styleguide do airbnb, e segundo pq as linhas ficam muito longas com 4 espaços e isso dificulta a leitura pra mim...

**Comment by lussolli** [comment]
Eu sempre uso quatro espaços com a tecla TAB, acho que assim fica mais fácil de ler o código no editor. O pessoal que trabalha comigo também faz isso.

**Comment by ninetails** [comment]
Uso 2 espaços, comecei depois de terem linters (airbnb) e o JS transpilado.

Antigamente (muito antigamente): Usava 4 para PHP (devido a algum padrão de mercado também mas nem lembro de qual config do PHPCS) e Tab pra JS não transpilado com espaçamento de 4 espaços.

**Comment by luizbills** [comment]
Eu uso tab com tamanho de 4 espaços para PHP (por ser um padrão da comunidade do WordPress). Mas fora do PHP (no HTML, JS e CSS) eu uso 4 espaços. Pessoalmente eu prefiro espaços. Mas não tenho problemas/frescuras nenhum em identar com tab. Tenho problemas é com código sem indentação ou identação zuada (misturando espaços e tab, por exemplo) :laughing:

**Comment by viniciusllima** [comment]
2. 
4 espaços parece Notepad++ na década de 90. Peguei ranço de 4 espaços. Haha

**Comment by felipemrodrigues** [comment]
2 espaços preferencialmente. Mas, acabo usando o que está configurado no projeto.

**Comment by klarkc** [comment]
whatever, uso o editorconfig :rofl:

**Comment by herberthenrique** [comment]
2 espaços :)

Com `.editorconfig` e `eslint` para garantir que todo o time está seguindo o mesmo guideline. :)

Engraçado que já usei quatro e achava super ok, depois que mudei pra 2 eu não quero mais voltar hahaha

**Comment by houstondapaz** [comment]
Para Javscript/HTML/Css - 2 espaços
Para C# - Tab

**Comment by drianoaz** [comment]
[Há algum tempo abri uma issue #953](https://github.com/frontendbr/forum/issues/953) para perguntar sobre regras de indentação, só que para  arquivos HTML. Nessa issue, o @ninetails disse uma coisa interessante:

> O ideal é utilizar um styleguide de mercado, pq aí quem pegar seu código já tem um padrão a seguir. 

O @kivervinicius também deu uma resposta bem interessante, que eu nunca havia percebido antes de ler isso: 

> a vantagem do espaço que ele sempre terá o mesmo tamanho em qualquer editor que você abrir, já tabs podem variar o tamanho da tabulação muitas vezes atrapalhando na leitura.

Junto a isso, também aprendi que os **tabs podem variar, não somente por um Editor/IDE, mas principalmente por Sistema Operacional.**

Até então, **eu utilizava 2 tabs** _(isso mesmo, não era espaços)_ para indentar meu código e não seguia nenhum styleguide _(simplesmente por não saber que isso existia)_. Não tenho vergonha de dizer que após abrir àquela issue despretenciosa, aprendi muita coisa bacana, por exemplo, como utilizar `.editorconfig`, `eslint` e `phplint` e outras ferramentas de qualidade de código. Desde então, meu código melhorou infinitamente.

Eu não conseguia entender como as pessoas conseguiam utilizar 2 espaços, principalmente no HTML, porque parecia que o código ficava sujo, dava a impressão de que estava tudo muito junto, mas é exatamente como o próprio @LFeh disse:

> Quando comecei a usar dessa forma _(2 espaços)_ (bons 5 anos atrás) achei estranho, dava a impressão da legibilidade ter diminuido, mas foi só questão de acostumar.

E realmente, tenho que concordar contigo @LFeh. Mas agora é exatamente o contrário: quando vejo um código com 4 espaços ao invés de 2, da vontade sair refatorando TUDO! HAHAHA Quando utilizo 2 espaços, tenho a impressão que fica mais fácil de ler meu código. Agora fico bem incomodado com linhas muito espaçadas.

**Eu particularmente, se pudesse utilizaria somente 2 espaços**, mas eu também trabalho com PHP e as [PSR's recomendam que é uma boa prática utilizarmos 4 espaços.](https://www.php-fig.org/psr/psr-2/#1-overview)

Basicamente sigo o seguinte padrão:

- 4 espaços para PHP e [blade templates do laravel](https://laravel.com/docs/5.6/blade)
- 2 espaços para o restante (HTML, CSS, JS), inclusive markdown.

Mais importante que tudo isso, aprendi que **devemos sempre manter a consistência do nosso código**, principalmente quando estamos trabalhando em equipe. Uma vez li, não sei onde, que  quando estamos trabalhando em conjunto com outros desenvolvedores em um mesmo projeto, **devemos fazer com que o código pareça ser escrito por apenas uma pessoa.**

Desde então tenho o costume de configurar o `.editorconfig`, `eslint` e `phplint` para garantir que não estou fugindo do padrão que adotamos aqui na empresa.

**Comment by felipeuntill** [comment]
Tab para C#/Swift/Etc ou 4 espaços para JS/CSS/HTML desde 2002 👴

E pra quem argumentar dizendo que a codebase fica mais leve eu digo, trabalhe e compre um computador / internet melhor  👍

**Comment by waghcwb** [comment]
Para Javascript, CSS, HTML, etc gosto de 2 espaços, acho que fica mais fácil de ler o código.
Para Python, Java, Kotlin e afins gosto de 4 espaços.

**Comment by ninetails** [comment]
Dois pontos que levo como verdades (pessoais) sobre o assunto de padrões de código:

- Democracia não é ideal para desenvolvimento web
- Padrão de mercado > padrão da empresa

As duas coisas andam juntas. Padrões de mercado atualizam e evoluem, padrões da empresa nem sempre atualizam na velocidade que evoluem padrões de mercado ou empresa grande que contribui com open source.

**Comment by diessica** [comment]
se isso é considerado polêmico, a gente tá mal de tópico........

**Comment by kazzkiq** [comment]
4 espaços é overkill. Sejamos minimalistas. 2 espaços *for the win*. 🙏

**Comment by julioc98** [comment]
8 espaços

**Comment by Adriano-Web** [comment]
Utilizo 4 espaços, pois acostumado com o Notepad++ editor que pré-define a indentação dos espaços como sendo 4. Mais como sou iniciante vocês devem entender. :/

**Comment by andreyroberto** [comment]
Eu vim parar aqui justamente tentando descobrir a melhor maneira de identar meu código em Java, sou iniciante mas não gosto de usar TAB e costumo utilizar 4 espaços, mas pelo visto a maioria do pessoal aqui utiliza 2, fiz uns testes aqui e não sei se consigo acostumar.

**Comment by saturnogallo** [comment]
> Eu vim parar aqui justamente tentando descobrir a melhor maneira de identar meu código em Java, sou iniciante mas não gosto de usar TAB e costumo utilizar 4 espaços, mas pelo visto a maioria do pessoal aqui utiliza 2, fiz uns testes aqui e não sei se consigo acostumar.

Eu pensava que não ia acostumar mas depois de escrever o primeiro código eu já estava acostumado. (Uso 2 espaços).

**Comment by gheysiell** [comment]
3 espaços

**Comment by felquis** [comment]
4 espaços

**Comment by thalesms2** [comment]
Eu uso conforme a IDE, tenho preferência por 4 espaços, mas várias vezes acontece da configuração padrão estar em 2 espaços e a preguiça para procurar como altera é maior que a necessidade de alterar a configuração.

**Comment by RatoX** [comment]
Eu tow usando setenta-e-cinco espaços. Vamos exercitar esse scrolling movement!

**Comment by felquis** [comment]
Setenta e quatro aqui

**Comment by ghost** [comment]
2 espaços.

**Comment by felquis** [comment]
Cada vez que aparecer um novo comentário aqui, vou aumento 1 espaço na minha indentação: 

São setenta e seis espaços que eu uso, quem da mais?
